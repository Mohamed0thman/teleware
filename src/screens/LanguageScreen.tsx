import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  I18nManager,
} from 'react-native';
import React, {useEffect} from 'react';
import {useAppDispatch, useAppSelector} from '../store/configureStore';
import {useTranslation} from 'react-i18next';
import RootScreen from '../component/RootScreen';
import {COLORS, FONTS, ICONS, SCALE} from '../constants';
import {StackScreenProps} from '@react-navigation/stack';
import {RootStackParamList} from '../navigation/RootNavigation';
import {checkAndRequestNotificationPermission} from '../utils/permissions';
import {changeLang, setNotifee} from '../store/slices/settingSlice';
import {Typography} from '../component/UI';

const {rw, rh, SC_Width} = SCALE;
const {EgyptFlag, AmircaFlag} = ICONS;

type Props = StackScreenProps<RootStackParamList, 'Language'> & {};

const LanguageScreen = ({navigation}: Props) => {
  const {t, i18n} = useTranslation();
  const {isRtl} = useAppSelector(state => state.setting);
  const dispatch = useAppDispatch();

  const handleOnSelectLang = (lang: 'ar' | 'en', isRtl: boolean) => {
    i18n.changeLanguage(lang);
    dispatch(changeLang({lang, isRtl}));
    navigation.replace('Location');
  };

  useEffect(() => {
    checkAndRequestNotificationPermission().then(() =>
      dispatch(setNotifee(true)),
    );
  }, []);

  //   render button to select lang
  const langSelector = (lang: 'ar' | 'en', title: string, isRtl: boolean) => {
    return (
      <TouchableOpacity
        testID={`${lang}-btn`}
        style={styles.langButton}
        onPress={() => handleOnSelectLang(lang, isRtl)}>
        {/* icon */}
        {lang === 'ar' ? (
          <EgyptFlag width={rw(50)} height={rh(50)} />
        ) : (
          <AmircaFlag width={rw(50)} height={rh(50)} />
        )}
        <Text style={styles.langTitle}>{title}</Text>
      </TouchableOpacity>
    );
  };

  return (
    <RootScreen containerStyle={styles.rootScreen}>
      <Typography testID="screen-title" isRtl={isRtl} style={styles.title}>
        {t('app:titles.langScreen')}
      </Typography>
      <View
        style={[styles.btns, {flexDirection: isRtl ? 'row-reverse' : 'row'}]}>
        {langSelector('en', 'English', false)}
        {langSelector('ar', 'العربيه', true)}
      </View>
    </RootScreen>
  );
};

export default LanguageScreen;

const styles = StyleSheet.create({
  rootScreen: {},
  title: {
    ...FONTS.h2,
    color: COLORS.black,
    textAlign: 'center',
    marginTop: rh(40),
  },
  btns: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    gap: 50,
  },
  langButton: {
    justifyContent: 'center',
    alignItems: 'center',
    width: rw(80),
    height: rh(120),
    borderColor: COLORS.black,
    borderWidth: 1,
  },
  langTitle: {...FONTS.h3, color: COLORS.black},
});
