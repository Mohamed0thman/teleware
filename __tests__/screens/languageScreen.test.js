import React from 'react';
import LanguageScreen from '../../src/screens/LanguageScreen';

import {fireEvent} from '@testing-library/react-native';
import {act} from 'react-test-renderer';

describe('languge screen', () => {
  it('renders correctly', () => {
    renderWithRedux(<LanguageScreen />);
  });

  it('renders correctly', () => {
    expect(renderWithRedux(<LanguageScreen />)).toMatchSnapshot();
  });

  it('should navigate to Location', async () => {
    // Wait for the component to be rendered
    const mockNavigation = {replace: jest.fn()};

    const {getByTestId} = renderWithRedux(
      <LanguageScreen navigation={mockNavigation} />,
    );

    fireEvent.press(getByTestId('ar-btn'));

    expect(mockNavigation.replace).toHaveBeenCalledTimes(1);
    expect(mockNavigation.replace).toHaveBeenCalledWith('Location');
  });
});
