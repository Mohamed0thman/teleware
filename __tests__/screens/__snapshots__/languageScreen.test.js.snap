// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`languge screen renders correctly 1`] = `
<View
  style={
    [
      {
        "paddingBottom": 0,
        "paddingTop": 0,
      },
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      },
      {},
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        {
          "textAlign": "left",
        },
        {
          "color": "#000",
          "fontFamily": "Roboto-Bold",
          "fontSize": 22,
          "marginTop": 65.5,
          "textAlign": "center",
        },
      ]
    }
    testID="screen-title"
  >
    app:titles.langScreen
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "gap": 50,
          "justifyContent": "center",
        },
        {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderColor": "#000",
          "borderWidth": 1,
          "height": 197,
          "justifyContent": "center",
          "opacity": 1,
          "width": 160,
        }
      }
      testID="en-btn"
    >
      <SvgMock
        height={82}
        width={100}
      />
      <Text
        style={
          {
            "color": "#000",
            "fontFamily": "Roboto-Bold",
            "fontSize": 20,
          }
        }
      >
        English
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderColor": "#000",
          "borderWidth": 1,
          "height": 197,
          "justifyContent": "center",
          "opacity": 1,
          "width": 160,
        }
      }
      testID="ar-btn"
    >
      <SvgMock
        height={82}
        width={100}
      />
      <Text
        style={
          {
            "color": "#000",
            "fontFamily": "Roboto-Bold",
            "fontSize": 20,
          }
        }
      >
        العربيه
      </Text>
    </View>
  </View>
</View>
`;
